File: packages/wire/pairing.go
package wire

import (
	"bytes"
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
)

// Pairing message types and domain separation constants.
const (
	MsgPairingRequest  = "pairing_request"
	MsgPairingResponse = "pairing_response"
	MsgPairingConfirm  = "pairing_confirm"

	DomainPairingRequest  = "sendbeam/1 pairing-request:"
	DomainPairingResponse = "sendbeam/1 pairing-response:"
	DomainPairingConfirm  = "sendbeam/1 pairing-confirm:"
	InfoPairCredential    = "sendbeam/1 pair-credential"

	PairingNonceSize = 32
)

var (
	// ErrInvalidPairingMessage is returned when a pairing message cannot be decoded or is malformed.
	ErrInvalidPairingMessage = errors.New("invalid pairing message")

	// ErrPairingSignatureFailed is returned when an Ed25519 device signature verification fails.
	ErrPairingSignatureFailed = errors.New("pairing signature verification failed")

	// ErrPairingDeviceIDMismatch is returned when the claimed device ID does not match the public key.
	ErrPairingDeviceIDMismatch = errors.New("pairing device ID does not match public key")

	// ErrPairingConfirmFailed is returned when the pairing confirmation authentication tag is invalid.
	ErrPairingConfirmFailed = errors.New("pairing confirmation authentication tag verification failed")

	// ErrPairingRejected is returned when a peer rejects the pairing ceremony.
	ErrPairingRejected = errors.New("pairing ceremony was rejected by peer")
)

// PairingRequest is sent by the initiating device in the pairing ceremony.
type PairingRequest struct {
	Type            string   `json:"type"`
	ProtocolVersion string   `json:"protocol_version"`
	DeviceID        string   `json:"device_id"`
	PublicKey       string   `json:"public_key"`
	DeviceName      string   `json:"device_name"`
	Capabilities    []string `json:"capabilities"`
	Nonce           string   `json:"nonce"`
	Signature       string   `json:"signature"`
}

// PairingResponse is sent by the receiving device upon verifying the PairingRequest.
type PairingResponse struct {
	Type            string   `json:"type"`
	ProtocolVersion string   `json:"protocol_version"`
	DeviceID        string   `json:"device_id"`
	PublicKey       string   `json:"public_key"`
	DeviceName      string   `json:"device_name"`
	Capabilities    []string `json:"capabilities"`
	Nonce           string   `json:"nonce"`
	Signature       string   `json:"signature"`
}

// PairingConfirm commits and finalizes the pairing ceremony after both sides agree to trust.
type PairingConfirm struct {
	Type    string `json:"type"`
	Status  string `json:"status"` // "accepted" or "rejected"
	AuthTag string `json:"auth_tag,omitempty"`
}

// BuildPairingRequestChallenge constructs the domain-separated message signed by the pairing initiator.
func BuildPairingRequestChallenge(masterKey, reqNonce []byte, deviceID string) []byte {
	h := sha256.Sum256(masterKey)
	buf := make([]byte, 0, len(DomainPairingRequest)+len(h)+len(reqNonce)+len(deviceID))
	buf = append(buf, DomainPairingRequest...)
	buf = append(buf, h[:]...)
	buf = append(buf, reqNonce...)
	buf = append(buf, deviceID...)
	return buf
}

// BuildPairingResponseChallenge constructs the domain-separated message signed by the pairing responder.
func BuildPairingResponseChallenge(masterKey, reqNonce, respNonce []byte, deviceID string) []byte {
	h := sha256.Sum256(masterKey)
	buf := make([]byte, 0, len(DomainPairingResponse)+len(h)+len(reqNonce)+len(respNonce)+len(deviceID))
	buf = append(buf, DomainPairingResponse...)
	buf = append(buf, h[:]...)
	buf = append(buf, reqNonce...)
	buf = append(buf, respNonce...)
	buf = append(buf, deviceID...)
	return buf
}

// DerivePairCredential derives the persistent pairwise credential (k_pair) and its reference ID.
// Public keys are sorted lexicographically to guarantee deterministic derivation regardless of role.
func DerivePairCredential(masterKey, reqNonce, respNonce, pubA, pubB []byte) ([]byte, string, error) {
	if len(masterKey) == 0 {
		return nil, "", errors.New("master key required")
	}
	if len(reqNonce) != PairingNonceSize || len(respNonce) != PairingNonceSize {
		return nil, "", errors.New("invalid pairing nonce length")
	}
	if len(pubA) != ed25519.PublicKeySize || len(pubB) != ed25519.PublicKeySize {
		return nil, "", errors.New("invalid public key size")
	}

	salt := make([]byte, 0, len(reqNonce)+len(respNonce))
	salt = append(salt, reqNonce...)
	salt = append(salt, respNonce...)

	var infoPubs []byte
	if bytes.Compare(pubA, pubB) < 0 {
		infoPubs = append(append([]byte(nil), pubA...), pubB...)
	} else {
		infoPubs = append(append([]byte(nil), pubB...), pubA...)
	}

	info := append([]byte(InfoPairCredential), infoPubs...)
	kPair, err := hkdfSHA256(masterKey, salt, info, 32)
	if err != nil {
		return nil, "", fmt.Errorf("derive k_pair: %w", err)
	}

	h := sha256.Sum256(kPair)
	credRef := "cred-" + hex.EncodeToString(h[:])

	return kPair, credRef, nil
}

// ComputePairingConfirmTag computes the HMAC-SHA256 confirmation tag over the peer's device ID.
func ComputePairingConfirmTag(kPair []byte, peerDeviceID string) string {
	mac := hmac.New(sha256.New, kPair)
	mac.Write([]byte(DomainPairingConfirm))
	mac.Write([]byte(peerDeviceID))
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifyPairingConfirmTag verifies the confirmation authentication tag using constant-time comparison.
func VerifyPairingConfirmTag(kPair []byte, peerDeviceID, tagHex string) bool {
	tagBytes, err := hex.DecodeString(tagHex)
	if err != nil || len(tagBytes) != sha256.Size {
		return false
	}
	expectedHex := ComputePairingConfirmTag(kPair, peerDeviceID)
	expectedBytes, _ := hex.DecodeString(expectedHex)
	return subtle.ConstantTimeCompare(tagBytes, expectedBytes) == 1
}

// NewPairingRequest creates a signed PairingRequest message.
func NewPairingRequest(id *DeviceIdentity, deviceName string, caps []string, masterKey []byte, nonce []byte) (*PairingRequest, error) {
	if id == nil {
		return nil, ErrInvalidIdentity
	}
	if len(nonce) != PairingNonceSize {
		nonce = make([]byte, PairingNonceSize)
		if _, err := rand.Read(nonce); err != nil {
			return nil, err
		}
	}

	challenge := BuildPairingRequestChallenge(masterKey, nonce, id.DeviceID)
	sig, err := id.Sign(challenge)
	if err != nil {
		return nil, fmt.Errorf("sign pairing request: %w", err)
	}

	return &PairingRequest{
		Type:            MsgPairingRequest,
		ProtocolVersion: "sendbeam/1",
		DeviceID:        id.DeviceID,
		PublicKey:       hex.EncodeToString(id.PublicKey),
		DeviceName:      deviceName,
		Capabilities:    caps,
		Nonce:           hex.EncodeToString(nonce),
		Signature:       hex.EncodeToString(sig),
	}, nil
}

// VerifyPairingRequest validates the format, DeviceID binding, and Ed25519 signature of a PairingRequest.
func VerifyPairingRequest(req *PairingRequest, masterKey []byte) (ed25519.PublicKey, []byte, error) {
	if req == nil || req.Type != MsgPairingRequest {
		return nil, nil, ErrInvalidPairingMessage
	}
	if !ValidateDeviceID(req.DeviceID) {
		return nil, nil, ErrInvalidDeviceID
	}

	pubBytes, err := hex.DecodeString(req.PublicKey)
	if err != nil || len(pubBytes) != ed25519.PublicKeySize {
		return nil, nil, ErrInvalidPublicKey
	}

	expectedID := DeriveDeviceID(pubBytes)
	if expectedID != req.DeviceID {
		return nil, nil, ErrPairingDeviceIDMismatch
	}

	nonceBytes, err := hex.DecodeString(req.Nonce)
	if err != nil || len(nonceBytes) != PairingNonceSize {
		return nil, nil, ErrInvalidPairingMessage
	}

	sigBytes, err := hex.DecodeString(req.Signature)
	if err != nil || len(sigBytes) != ed25519.SignatureSize {
		return nil, nil, ErrPairingSignatureFailed
	}

	challenge := BuildPairingRequestChallenge(masterKey, nonceBytes, req.DeviceID)
	if !VerifyDeviceSignature(pubBytes, challenge, sigBytes) {
		return nil, nil, ErrPairingSignatureFailed
	}

	return pubBytes, nonceBytes, nil
}

// NewPairingResponse creates a signed PairingResponse message responding to a verified request.
func NewPairingResponse(id *DeviceIdentity, deviceName string, caps []string, masterKey []byte, reqNonce, respNonce []byte) (*PairingResponse, error) {
	if id == nil {
		return nil, ErrInvalidIdentity
	}
	if len(reqNonce) != PairingNonceSize {
		return nil, errors.New("invalid request nonce length")
	}
	if len(respNonce) != PairingNonceSize {
		respNonce = make([]byte, PairingNonceSize)
		if _, err := rand.Read(respNonce); err != nil {
			return nil, err
		}
	}

	challenge := BuildPairingResponseChallenge(masterKey, reqNonce, respNonce, id.DeviceID)
	sig, err := id.Sign(challenge)
	if err != nil {
		return nil, fmt.Errorf("sign pairing response: %w", err)
	}

	return &PairingResponse{
		Type:            MsgPairingResponse,
		ProtocolVersion: "sendbeam/1",
		DeviceID:        id.DeviceID,
		PublicKey:       hex.EncodeToString(id.PublicKey),
		DeviceName:      deviceName,
		Capabilities:    caps,
		Nonce:           hex.EncodeToString(respNonce),
		Signature:       hex.EncodeToString(sig),
	}, nil
}

// VerifyPairingResponse validates the format, DeviceID binding, and Ed25519 signature of a PairingResponse.
func VerifyPairingResponse(resp *PairingResponse, reqNonce []byte, masterKey []byte) (ed25519.PublicKey, []byte, error) {
	if resp == nil || resp.Type != MsgPairingResponse {
		return nil, nil, ErrInvalidPairingMessage
	}
	if !ValidateDeviceID(resp.DeviceID) {
		return nil, nil, ErrInvalidDeviceID
	}

	pubBytes, err := hex.DecodeString(resp.PublicKey)
	if err != nil || len(pubBytes) != ed25519.PublicKeySize {
		return nil, nil, ErrInvalidPublicKey
	}

	expectedID := DeriveDeviceID(pubBytes)
	if expectedID != resp.DeviceID {
		return nil, nil, ErrPairingDeviceIDMismatch
	}

	respNonceBytes, err := hex.DecodeString(resp.Nonce)
	if err != nil || len(respNonceBytes) != PairingNonceSize {
		return nil, nil, ErrInvalidPairingMessage
	}

	sigBytes, err := hex.DecodeString(resp.Signature)
	if err != nil || len(sigBytes) != ed25519.SignatureSize {
		return nil, nil, ErrPairingSignatureFailed
	}

	challenge := BuildPairingResponseChallenge(masterKey, reqNonce, respNonceBytes, resp.DeviceID)
	if !VerifyDeviceSignature(pubBytes, challenge, sigBytes) {
		return nil, nil, ErrPairingSignatureFailed
	}

	return pubBytes, respNonceBytes, nil
}

// NewPairingConfirm creates a PairingConfirm commit message.
func NewPairingConfirm(kPair []byte, peerDeviceID string, accepted bool) *PairingConfirm {
	if !accepted {
		return &PairingConfirm{
			Type:   MsgPairingConfirm,
			Status: "rejected",
		}
	}
	tag := ComputePairingConfirmTag(kPair, peerDeviceID)
	return &PairingConfirm{
		Type:    MsgPairingConfirm,
		Status:  "accepted",
		AuthTag: tag,
	}
}

// VerifyPairingConfirm verifies an incoming PairingConfirm message.
func VerifyPairingConfirm(confirm *PairingConfirm, kPair []byte, peerDeviceID string) error {
	if confirm == nil || confirm.Type != MsgPairingConfirm {
		return ErrInvalidPairingMessage
	}
	if confirm.Status != "accepted" {
		return ErrPairingRejected
	}
	if !VerifyPairingConfirmTag(kPair, peerDeviceID, confirm.AuthTag) {
		return ErrPairingConfirmFailed
	}
	return nil
}

// EncodePairingMessage marshals any pairing message into compact JSON bytes.
func EncodePairingMessage(msg any) ([]byte, error) {
	return json.Marshal(msg)
}

// DecodePairingMessage unmarshals a pairing message into its concrete type.
func DecodePairingMessage(data []byte) (any, error) {
	var peek struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(data, &peek); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrInvalidPairingMessage, err)
	}

	switch peek.Type {
	case MsgPairingRequest:
		var req PairingRequest
		if err := json.Unmarshal(data, &req); err != nil {
			return nil, err
		}
		return &req, nil
	case MsgPairingResponse:
		var resp PairingResponse
		if err := json.Unmarshal(data, &resp); err != nil {
			return nil, err
		}
		return &resp, nil
	case MsgPairingConfirm:
		var conf PairingConfirm
		if err := json.Unmarshal(data, &conf); err != nil {
			return nil, err
		}
		return &conf, nil
	default:
		return nil, fmt.Errorf("%w: unknown type %q", ErrInvalidPairingMessage, peek.Type)
	}
}

File: packages/protocol/src/pairing.ts
/**
 * Pairing ceremony protocol messages, cryptographic challenges, and pair-specific
 * credential derivations for SendBeam (V15-PR02).
 *
 * Matches Go `packages/wire/pairing.go` byte-for-byte.
 */

import { bytesToHex, concatBytes, hexToBytes, utf8 } from './bytes.js';
import {
  deriveDeviceId,
  signDeviceMessage,
  validateDeviceId,
  verifyDeviceSignature,
  type DeviceIdentity,
} from './identity.js';
import { hkdfSha256, hmacSha256, randomBytes, sha256 } from './webcrypto.js';

export const MSG_PAIRING_REQUEST = 'pairing_request';
export const MSG_PAIRING_RESPONSE = 'pairing_response';
export const MSG_PAIRING_CONFIRM = 'pairing_confirm';

export const DOMAIN_PAIRING_REQUEST = 'sendbeam/1 pairing-request:';
export const DOMAIN_PAIRING_RESPONSE = 'sendbeam/1 pairing-response:';
export const DOMAIN_PAIRING_CONFIRM = 'sendbeam/1 pairing-confirm:';
export const INFO_PAIR_CREDENTIAL = 'sendbeam/1 pair-credential';

export const PAIRING_NONCE_BYTES = 32;

export interface PairingRequest {
  readonly type: typeof MSG_PAIRING_REQUEST;
  readonly protocol_version: string;
  readonly device_id: string;
  readonly public_key: string;
  readonly device_name: string;
  readonly capabilities: string[];
  readonly nonce: string;
  readonly signature: string;
}

export interface PairingResponse {
  readonly type: typeof MSG_PAIRING_RESPONSE;
  readonly protocol_version: string;
  readonly device_id: string;
  readonly public_key: string;
  readonly device_name: string;
  readonly capabilities: string[];
  readonly nonce: string;
  readonly signature: string;
}

export interface PairingConfirm {
  readonly type: typeof MSG_PAIRING_CONFIRM;
  readonly status: 'accepted' | 'rejected';
  readonly auth_tag?: string;
}

export type PairingMessage = PairingRequest | PairingResponse | PairingConfirm;

/**
 * Compare two byte arrays lexicographically.
 */
function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i]! < b[i]!) return -1;
    if (a[i]! > b[i]!) return 1;
  }
  return a.length - b.length;
}

/**
 * Constant-time hex string comparison.
 */
function constantTimeHexEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/**
 * Build the domain-separated message challenge for a PairingRequest.
 */
export async function buildPairingRequestChallenge(
  masterKey: Uint8Array,
  reqNonce: Uint8Array,
  deviceId: string,
): Promise<Uint8Array> {
  const masterHash = await sha256(masterKey);
  return concatBytes(utf8(DOMAIN_PAIRING_REQUEST), masterHash, reqNonce, utf8(deviceId));
}

/**
 * Build the domain-separated message challenge for a PairingResponse.
 */
export async function buildPairingResponseChallenge(
  masterKey: Uint8Array,
  reqNonce: Uint8Array,
  respNonce: Uint8Array,
  deviceId: string,
): Promise<Uint8Array> {
  const masterHash = await sha256(masterKey);
  return concatBytes(
    utf8(DOMAIN_PAIRING_RESPONSE),
    masterHash,
    reqNonce,
    respNonce,
    utf8(deviceId),
  );
}

/**
 * Derive the pairwise persistent credential (`kPair`) and its reference ID (`credRef`).
 * Public keys are sorted lexicographically to guarantee deterministic symmetry.
 */
export async function derivePairCredential(
  masterKey: Uint8Array,
  reqNonce: Uint8Array,
  respNonce: Uint8Array,
  pubA: Uint8Array,
  pubB: Uint8Array,
): Promise<{ kPair: Uint8Array; credRef: string }> {
  if (masterKey.length === 0) {
    throw new Error('master key required');
  }
  if (reqNonce.length !== PAIRING_NONCE_BYTES || respNonce.length !== PAIRING_NONCE_BYTES) {
    throw new Error('invalid pairing nonce length');
  }
  if (pubA.length !== 32 || pubB.length !== 32) {
    throw new Error('invalid public key size');
  }

  const salt = concatBytes(reqNonce, respNonce);
  const infoPubs = compareBytes(pubA, pubB) < 0 ? concatBytes(pubA, pubB) : concatBytes(pubB, pubA);
  const info = concatBytes(utf8(INFO_PAIR_CREDENTIAL), infoPubs);

  const kPair = await hkdfSha256(masterKey, salt, info, 32);
  const credHash = await sha256(kPair);
  const credRef = `cred-${bytesToHex(credHash)}`;

  return { kPair, credRef };
}

/**
 * Compute the confirmation HMAC-SHA256 authentication tag for the peer device.
 */
export async function computePairingConfirmTag(
  kPair: Uint8Array,
  peerDeviceId: string,
): Promise<string> {
  const data = concatBytes(utf8(DOMAIN_PAIRING_CONFIRM), utf8(peerDeviceId));
  const tag = await hmacSha256(kPair, data);
  return bytesToHex(tag);
}

/**
 * Verify the confirmation HMAC-SHA256 tag.
 */
export async function verifyPairingConfirmTag(
  kPair: Uint8Array,
  peerDeviceId: string,
  tagHex: string,
): Promise<boolean> {
  const expected = await computePairingConfirmTag(kPair, peerDeviceId);
  return constantTimeHexEqual(tagHex.toLowerCase(), expected.toLowerCase());
}

/**
 * Create a signed PairingRequest message.
 */
export async function createPairingRequest(
  id: DeviceIdentity,
  deviceName: string,
  capabilities: string[],
  masterKey: Uint8Array,
  nonce?: Uint8Array,
): Promise<PairingRequest> {
  const reqNonce =
    nonce && nonce.length === PAIRING_NONCE_BYTES ? nonce : randomBytes(PAIRING_NONCE_BYTES);
  const challenge = await buildPairingRequestChallenge(masterKey, reqNonce, id.deviceId);
  const sig = signDeviceMessage(id, challenge);

  return {
    type: MSG_PAIRING_REQUEST,
    protocol_version: 'sendbeam/1',
    device_id: id.deviceId,
    public_key: bytesToHex(id.publicKey),
    device_name: deviceName,
    capabilities,
    nonce: bytesToHex(reqNonce),
    signature: bytesToHex(sig),
  };
}

/**
 * Verify a PairingRequest message format, DeviceID binding, and Ed25519 signature.
 */
export async function verifyPairingRequest(
  req: PairingRequest,
  masterKey: Uint8Array,
): Promise<{ publicKey: Uint8Array; nonce: Uint8Array }> {
  if (!req || req.type !== MSG_PAIRING_REQUEST) {
    throw new Error('invalid pairing message');
  }
  if (!validateDeviceId(req.device_id)) {
    throw new Error('invalid device id format');
  }

  const pubBytes = hexToBytes(req.public_key);
  if (pubBytes.length !== 32) {
    throw new Error('invalid public key: expected 32-byte Ed25519 public key');
  }

  const expectedId = await deriveDeviceId(pubBytes);
  if (expectedId !== req.device_id) {
    throw new Error('pairing device ID does not match public key');
  }

  const nonceBytes = hexToBytes(req.nonce);
  if (nonceBytes.length !== PAIRING_NONCE_BYTES) {
    throw new Error('invalid pairing nonce length');
  }

  const sigBytes = hexToBytes(req.signature);
  if (sigBytes.length !== 64) {
    throw new Error('pairing signature verification failed');
  }

  const challenge = await buildPairingRequestChallenge(masterKey, nonceBytes, req.device_id);
  const valid = verifyDeviceSignature(pubBytes, challenge, sigBytes);
  if (!valid) {
    throw new Error('pairing signature verification failed');
  }

  return { publicKey: pubBytes, nonce: nonceBytes };
}

/**
 * Create a signed PairingResponse message responding to a verified PairingRequest.
 */
export async function createPairingResponse(
  id: DeviceIdentity,
  deviceName: string,
  capabilities: string[],
  masterKey: Uint8Array,
  reqNonce: Uint8Array,
  respNonce?: Uint8Array,
): Promise<PairingResponse> {
  if (reqNonce.length !== PAIRING_NONCE_BYTES) {
    throw new Error('invalid request nonce length');
  }
  const nonce =
    respNonce && respNonce.length === PAIRING_NONCE_BYTES
      ? respNonce
      : randomBytes(PAIRING_NONCE_BYTES);
  const challenge = await buildPairingResponseChallenge(masterKey, reqNonce, nonce, id.deviceId);
  const sig = signDeviceMessage(id, challenge);

  return {
    type: MSG_PAIRING_RESPONSE,
    protocol_version: 'sendbeam/1',
    device_id: id.deviceId,
    public_key: bytesToHex(id.publicKey),
    device_name: deviceName,
    capabilities,
    nonce: bytesToHex(nonce),
    signature: bytesToHex(sig),
  };
}

/**
 * Verify a PairingResponse message format, DeviceID binding, and Ed25519 signature.
 */
export async function verifyPairingResponse(
  resp: PairingResponse,
  reqNonce: Uint8Array,
  masterKey: Uint8Array,
): Promise<{ publicKey: Uint8Array; nonce: Uint8Array }> {
  if (!resp || resp.type !== MSG_PAIRING_RESPONSE) {
    throw new Error('invalid pairing message');
  }
  if (!validateDeviceId(resp.device_id)) {
    throw new Error('invalid device id format');
  }

  const pubBytes = hexToBytes(resp.public_key);
  if (pubBytes.length !== 32) {
    throw new Error('invalid public key: expected 32-byte Ed25519 public key');
  }

  const expectedId = await deriveDeviceId(pubBytes);
  if (expectedId !== resp.device_id) {
    throw new Error('pairing device ID does not match public key');
  }

  const respNonceBytes = hexToBytes(resp.nonce);
  if (respNonceBytes.length !== PAIRING_NONCE_BYTES) {
    throw new Error('invalid pairing nonce length');
  }

  const sigBytes = hexToBytes(resp.signature);
  if (sigBytes.length !== 64) {
    throw new Error('pairing signature verification failed');
  }

  const challenge = await buildPairingResponseChallenge(
    masterKey,
    reqNonce,
    respNonceBytes,
    resp.device_id,
  );
  const valid = verifyDeviceSignature(pubBytes, challenge, sigBytes);
  if (!valid) {
    throw new Error('pairing signature verification failed');
  }

  return { publicKey: pubBytes, nonce: respNonceBytes };
}

/**
 * Create a PairingConfirm commit message.
 */
export async function createPairingConfirm(
  kPair: Uint8Array,
  peerDeviceId: string,
  accepted: boolean,
): Promise<PairingConfirm> {
  if (!accepted) {
    return {
      type: MSG_PAIRING_CONFIRM,
      status: 'rejected',
    };
  }
  const authTag = await computePairingConfirmTag(kPair, peerDeviceId);
  return {
    type: MSG_PAIRING_CONFIRM,
    status: 'accepted',
    auth_tag: authTag,
  };
}

/**
 * Verify a PairingConfirm commit message.
 */
export async function verifyPairingConfirm(
  confirm: PairingConfirm,
  kPair: Uint8Array,
  peerDeviceId: string,
): Promise<void> {
  if (!confirm || confirm.type !== MSG_PAIRING_CONFIRM) {
    throw new Error('invalid pairing message');
  }
  if (confirm.status !== 'accepted') {
    throw new Error('pairing ceremony was rejected by peer');
  }
  if (
    !confirm.auth_tag ||
    !(await verifyPairingConfirmTag(kPair, peerDeviceId, confirm.auth_tag))
  ) {
    throw new Error('pairing confirmation authentication tag verification failed');
  }
}
